src/components/MessageList/renderMessages.tsx

```diff
@@ -0,0 +1,107 @@
+import React, { ReactNode } from 'react';
+
+import { MessageProps } from '../Message/types';
+import { StreamMessage } from '../../context/ChannelStateContext';
+import { DefaultStreamChatGenerics } from '../../types/types';
+import { ComponentContextValue, CustomClasses, isDate } from '../../context';
+import { CUSTOM_MESSAGE_TYPE } from '../../constants/messageTypes';
+import { GroupStyle } from './utils';
+import { Message } from '../Message/Message';
+import { UserResponse } from 'stream-chat';
+
+export interface RenderMessagesOptions<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+> {
+  components: ComponentContextValue<StreamChatGenerics>;
+  lastReceivedId: string | null;
+  messageGroupStyles: Record<string, GroupStyle>;
+  messageProps: Omit<MessageProps<StreamChatGenerics>, MessagePropsToOmit>;
+  messages: Array<StreamMessage<StreamChatGenerics>>;
+  readData: Record<string, Array<UserResponse<StreamChatGenerics>>>;
+  threadList: boolean;
+  customClasses?: CustomClasses;
+}
+
+export type MessageRenderer<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+> = (options: RenderMessagesOptions<StreamChatGenerics>) => Array<ReactNode>;
+
+type MessagePropsToOmit =
+  | 'channel'
+  | 'groupStyles'
+  | 'initialMessage'
+  | 'lastReceivedId'
+  | 'message'
+  | 'readBy'
+  | 'threadList';
+
+export function defaultRenderMessages<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+>({
+  components,
+  customClasses,
+  lastReceivedId,
+  messageGroupStyles,
+  messageProps,
```

Comment:
As we are rendering a list of messages it could be unclear, for which the `messageProps` param is destined. Maybe we could give it another name like `sharedMessageProps`.

Reply:
Fixed in https://github.com/GetStream/stream-chat-react/pull/2243/commits/bd5a39997957c6c993add0446d5f744dbcdb0a2b.